event-web-framework/draft.markdown

```diff
@@ -0,0 +1,872 @@
+
+As far as I know, anaphoric macros were first defined by Paul Graham in an [OnLisp chapter](http://dunsmor.com/lisp/onlisp/onlisp_18.html). The use case he gives is a situation where you want to do some sort of expensive or semi-expensive check, then do something conditionally on the result. In the above context, we're using `aif` to do a check the result of an `alist` traversal.
+
+```lisp
+	(aif (cdr (assoc :room parameters))
+	     (uri-decode it)
+	     (error (make-instance 'http-assertion-error :assertion 'room)))
+```
+
+This takes the `cdr` of looking up the symbol `:room` in the association list `parameters`. If that returns a non-nil value, `uri-decode` it, otherwise throw an error of the type `http-assertion-error`.
+
+In other words, the above is equivalent to
+
+```lisp
+	(let ((it (cdr (assoc :room parameters))))
+	  (if it
+	      (uri-decode it)
+	      (error (make-instance 'http-assertion-error :assertion 'room))))
+```
+
+Strongly-typed functional languages like Haskell often use a `Maybe` type in this situation. In Common Lisp, we capture the symbol `it` in the expansion as the name for the result of the check. 
+
+Understanding this, we should be able to see that `arg-exp` is generating a specific, repetitive, piece of the code tree that we eventually want to evaluate. In this case, the piece that checks for the presence of the given parameter among the handlers' `parameters`. Now, let's move onto...
+
+#### type-expression
+
+```lisp
+	(defgeneric type-expression (parameter type)
+	  (:documentation
+	   "A type-expression will tell the server how to convert a parameter from a string to a particular, necessary type."))
+    ...
+	(defmethod type-expression (parameter type) nil)
+```
+
+This is a generic function that generates new tree structures (coincidentally Lisp code), rather than just a function. The only thing the above tells you is that by default, a `type-expression` is `NIL`. Which is to say, we don't have one. If we encounter a `NIL`, we use the raw output of `arg-exp`, but that doesn't tell us much about the most common case. To see that, lets take a look at a built-in (to `:house`) `define-http-type` expression.
+
+```lisp
+	(define-http-type (:integer)
+	    :type-expression `(parse-integer ,parameter :junk-allowed t)
+		:type-assertion `(numberp ,parameter))
+```
+
+An `:integer` is something we're making from a `parameter` by using `(parse-integer parameter :junk-allowed t)`. `junk-allowed` tells `parse-integer` that we're not confident the data we're giving it is actually parseable, so we need to make sure that the returned result is an integer. If it isn't, we get this behaviour: 
```

Comment:
Note to self: Look at this and the accompanying REPL block. Might just be a brain-fart on my part, but it looks like this might not make sense.